README.md
# code_split

Split Python code files per class and function.

## Description

This script will split a Python code file into separate files per main class or function.
The new files will be named as per the class or function it contains.
The newly created files will contain class/function decorators as well as comments directly before the class or function (without empty lines).

Intermediate comments and the main code will be ignored.

## Usage

The tool requires at least the source code file name, optionally an output path can be specified.
The source file name and the output folder can be relative to the current working directory or with absolute path.

```text
usage: code_split [-h] [--version] -i INPUT [-f FOLDER] [-v] [-vv]

Python code split tool

options:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  -i INPUT, --input INPUT
                        Python code file to be split
  -f FOLDER, --folder FOLDER
                        Destination folder for the split code
  -v, --verbose         set loglevel to INFO
  -vv, --very-verbose   set loglevel to DEBUG
```

### Example

Below shows the source files and the new files created from it by the command `code_split -i source_code.py`:

`source_code.py`

```python
"""Source header and imports will be ignored"""
import logging
from dataclasses import dataclass

logging.getLogger(__name__)

# This is the first exported class
@dataclass
class MyData:
    """Sample dataclass"""

    name: str
    age: int

    def birthday(self):
        self.age += 1


# Intermediate comments
# will be ignored

def my_function(data: MyData) -> str:
    dummy = SampleClass(data.name)
    # Sample comment
    return f"{dummy.output()}, age: {data.age}"

# Main code starts here
print("End of the example source code.")

```

The resulting files will be:

`MyData.py`

```python
# This is the first exported class
@dataclass
class MyData:
    """Sample dataclass"""

    name: str
    age: int

    def birthday(self):
        self.age += 1

```

`my_function.py`

```python
def my_function(data: MyData) -> str:
    dummy = SampleClass(data.name)
    # Sample comment
    return f"{dummy.output()}, age: {data.age}"

```

<!-- pyscaffold-notes -->

## Note

This project has been set up using PyScaffold 4.2.2. For details and usage
information on PyScaffold see <https://pyscaffold.org/>.


CHANGELOG.md
# Changelog

## Version 0.9.0 (RC1)

Date: 2022-06-26

- First release with minimum required features
- Full pytest coverage of the code


tests/conftest.py
"""
    Dummy conftest.py for code_split.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html
"""

# import pytest


tests/test_skeleton.py
import pytest

from code_split.skeleton import fib, main

__author__ = "Matthias Homann"
__copyright__ = "Matthias Homann"
__license__ = "GPL-3.0-or-later"


def test_fib():
    """API Tests"""
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


def test_main(capsys):
    """CLI Tests"""
    # capsys is a pytest fixture that allows asserts against stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    main(["7"])
    captured = capsys.readouterr()
    assert "The 7-th Fibonacci number is 13" in captured.out


tests/test_code_split.py
import logging
import os
import sys
from pathlib import Path

import pytest
from fixtures.sample_data import code

from code_split import __version__
from code_split.code_split import main, run

__author__ = "Matthias Homann"
__copyright__ = "Matthias Homann"
__license__ = "GPL-3.0-or-later"


def test_main_version(capsys):
    """Basic test of main()"""
    # capsys is a pytest fixture that allows asserts against stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    with pytest.raises(SystemExit) as pytest_exit:
        main(["--version"])
    captured = capsys.readouterr()
    assert __version__ in captured.out
    assert pytest_exit.type == SystemExit
    assert pytest_exit.value.code == 0


def test_run_version(capsys, monkeypatch):
    """CLI Tests of run()"""
    with pytest.raises(SystemExit) as pytest_exit:
        with monkeypatch.context() as m:
            m.setattr(sys, "argv", ["code_split", "--version"])
            run()
    captured = capsys.readouterr()
    assert __version__ in captured.out
    assert pytest_exit.type == SystemExit
    assert pytest_exit.value.code == 0


def test_main_input_missing(capsys):
    """CLI Tests"""
    # capsys is a pytest fixture that allows asserts against stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    with pytest.raises(SystemExit) as pytest_exit:
        main([])
    captured = capsys.readouterr()
    assert "required: -i/--input" in captured.err
    assert pytest_exit.type == SystemExit
    assert pytest_exit.value.code == 2


def test_code_split_abs_path(tmp_path):
    """Test code_split with sample date using absolute path

    First the sample code is created in the folder "tmp_path/source/" from the sample_data.code
    and then the code_split is applies and the result compared with the sample_data snippets.

    Parameters
    ----------
    tmp_path : Path
        Temp path fixture
    """
    d = tmp_path / "source"
    d.mkdir()
    src = d / "test_code.py"
    src.write_text("".join(code.values()))
    assert src.read_text() == "".join(code.values())
    main(["-i", str(src), "-f", str(d)])
    for name, value in code.items():
        if not name.startswith("skip"):
            my_data = d / f"{name}.py"
            assert my_data.read_text() == value


def test_code_split_rel_path(tmp_path):
    """Test code_split with sample date using relative path

    First the sample code is created in the folder "tmp_path/source/" from the sample_data.code
    and then the code_split is applies and the result compared with the sample_data snippets.

    Parameters
    ----------
    tmp_path : Path
        Temp path fixture
    """
    # Change working directory to tmp_path
    os.chdir(tmp_path)
    assert str(tmp_path) == os.getcwd()
    d = tmp_path / "source"
    d.mkdir()
    src = d / "test_code.py"
    src.write_text("".join(code.values()))
    assert src.read_text() == "".join(code.values())
    main(["-i", "source/test_code.py", "-f", "source"])
    for name, value in code.items():
        if not name.startswith("skip"):
            my_data = d / f"{name}.py"
            assert my_data.read_text() == value


def test_code_split_no_path(tmp_path):
    """Test code_split with sample date without specifying the path

    First the sample code is created in the folder "tmp_path/source/" from the sample_data.code
    and then the code_split is applies and the result compared with the sample_data snippets.

    Parameters
    ----------
    tmp_path : Path
        Temp path fixture
    """
    # Change working directory to tmp_path/source
    d = tmp_path / "source"
    d.mkdir()
    os.chdir(tmp_path / "source")
    assert str(tmp_path / "source") == os.getcwd()
    src = d / "test_code.py"
    src.write_text("".join(code.values()))
    assert src.read_text() == "".join(code.values())
    main(["-i", "test_code.py"])
    for name, value in code.items():
        if not name.startswith("skip"):
            my_data = d / f"{name}.py"
            assert my_data.read_text() == value


def test_code_split_new_folder(tmp_path):
    """Test code_split with sample date using relative path

    First the sample code is created in the folder "tmp_path/source/" from the sample_data.code
    and then the code_split is applies and the result compared with the sample_data snippets.

    Parameters
    ----------
    tmp_path : Path
        Temp path fixture
    """
    # Change working directory to tmp_path/source
    d = tmp_path / "source"
    d.mkdir()
    os.chdir(tmp_path / "source")
    assert str(tmp_path / "source") == os.getcwd()
    src = d / "test_code.py"
    src.write_text("".join(code.values()))
    assert src.read_text() == "".join(code.values())
    main(["-i", "test_code.py", "-f", "new"])
    for name, value in code.items():
        if not name.startswith("skip"):
            my_data = d / "new" / f"{name}.py"
            assert my_data.read_text() == value


def test_code_split_file_not_found(caplog, tmp_path):
    """Test code_split with non-existing input file

    Parameters
    ----------
    caplog : fixture
    tmp_path : Path
        Temp path fixture
    """
    caplog.set_level(logging.ERROR)
    d = tmp_path / "source"
    d.mkdir()
    src = d / "test_code_not_found.py"
    main(["-i", str(src)])
    assert caplog.record_tuples == [
        (
            "code_split.code_split",
            logging.ERROR,
            f"Can't find input file {src}",
        )
    ]


def test_code_split_single_def(tmp_path):
    """Test code_split with sample date which has no main

    First the sample code is created in the folder "tmp_path/source/" from first function of sample_data.code
    and then the code_split is applies and the result compared with the sample_data snippets.

    Parameters
    ----------
    tmp_path : Path
        Temp path fixture
    """
    d = tmp_path / "source"
    d.mkdir()
    src = d / "test_code.py"
    result = ""
    for name, value in code.items():
        if value.startswith("def"):
            result = value
            src.write_text(value)
            break
    assert src.read_text() == result
    main(["-i", str(src), "-f", str(d)])
    for name, value in code.items():
        if name.startswith("def"):
            my_data = d / f"{name}.py"
            assert my_data.read_text() == value
            break


src/code_split/skeleton.py
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following lines in the
``[options.entry_points]`` section in ``setup.cfg``::

    console_scripts =
         fibonacci = code_split.skeleton:run

Then run ``pip install .`` (or ``pip install -e .`` for editable mode)
which will install the command ``fibonacci`` inside your current environment.

Besides console scripts, the header (i.e. until ``_logger``...) of this file can
also be used as template for Python modules.

Note:
    This file can be renamed depending on your needs or safely removed if not needed.

References:
    - https://setuptools.pypa.io/en/latest/userguide/entry_point.html
    - https://pip.pypa.io/en/stable/reference/pip_install
"""

import argparse
import logging
import sys

from code_split import __version__

__author__ = "Matthias Homann"
__copyright__ = "Matthias Homann"
__license__ = "GPL-3.0-or-later"

_logger = logging.getLogger(__name__)


# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from code_split.skeleton import fib`,
# when using this Python module as a library.


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for _i in range(n - 1):
        a, b = b, a + b
    return a


# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.


def parse_args(args):
    """Parse command line parameters

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--help"]``).

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(description="Just a Fibonacci demonstration")
    parser.add_argument(
        "--version",
        action="version",
        version="code_split {ver}".format(ver=__version__),
    )
    parser.add_argument(dest="n", help="n-th Fibonacci number", type=int, metavar="INT")
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel, stream=sys.stdout, format=logformat, datefmt="%Y-%m-%d %H:%M:%S"
    )


def main(args):
    """Wrapper allowing :func:`fib` to be called with string arguments in a CLI fashion

    Instead of returning the value from :func:`fib`, it prints the result to the
    ``stdout`` in a nicely formatted message.

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--verbose", "42"]``).
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m code_split.skeleton 42
    #
    run()


src/code_split/code_split.py
"""
Split Python code files per class and function
"""

import argparse
import logging
import re
import sys
from pathlib import Path
from typing import List

from attr import s

from code_split import __version__

__author__ = "Matthias Homann"
__copyright__ = "Matthias Homann"
__license__ = "GPL-3.0-or-later"

_logger = logging.getLogger(__name__)


# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from code_split.code_split import split_code`,
# when using this Python module as a library.


def split_code(src_code: str, folder: str) -> None:
    """Reads the source code file and writes a new output file
    per contained top level class and function.

    The functions accepts relative and absolute paths for the scr_code and folder.

    Parameters
    ----------
    src_code : str
        Name of the source code which will be used as input
    folder : str
        Output folder for the new files
    """
    src_path = Path(src_code)
    if not src_path.is_absolute():
        src_path = Path.cwd().joinpath(src_path)
        _logger.debug("Appended CWD to input file path")
    if folder:
        output = Path(folder)
        if not output.is_absolute():
            output = Path.cwd().joinpath(output)
    else:
        output = Path.cwd()

    if not output.is_dir():
        _logger.info("Create output folder %s", output)
        output.mkdir()
    out_file = None
    cache = ""
    blank_lines = ""
    pre_comment = ""
    try:
        with src_path.open(encoding="utf-8") as file:
            for line in file.readlines():
                if line.startswith("@"):
                    cache += line
                if line.startswith("def") or line.startswith("class"):
                    out_file_name = re.findall(r"^\w+\s+(\w+).*", line)[0] + ".py"
                    _logger.info("NEW output file: %s", out_file_name)
                    if out_file:
                        out_file.close()
                    out_file = output.joinpath(out_file_name).open("w", encoding="utf-8")
                    out_file.write(pre_comment)
                    pre_comment = ""
                    out_file.write(cache)
                    cache = ""
                    blank_lines = ""
                elif not line.startswith(" ") and len(line.strip()) and out_file:
                    # Class of function ended, either comments or main code
                    out_file.close()
                    out_file = None

                if not line.strip():
                    # cache blank lines
                    blank_lines += line
                    # ignore comments before functions is separated by a blank line
                    pre_comment = ""
                elif out_file:
                    if blank_lines:
                        out_file.write(blank_lines)
                        blank_lines = ""
                    _logger.debug("> %s", line.strip())
                    out_file.write(line)
                elif line.startswith("#"):
                    pre_comment += line
        if out_file:
            out_file.close()
    except FileNotFoundError:
        _logger.error("Can't find input file %s", src_code)


# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.


def parse_args(args: List[str]) -> argparse.Namespace:
    """Parse command line parameters

    Parameters
    ----------
    args : List[str]
        command line parameters as list of strings
        (for example  ``["--help"]``).

    Returns
    -------
    argparse.Namespace
        command line parameters namespace
    """
    parser = argparse.ArgumentParser(description="Python code split tool")
    parser.add_argument(
        "--version",
        action="version",
        version="code_split {ver}".format(ver=__version__),
    )
    parser.add_argument("-i", "--input", required=True, type=str, help="Python code file to be split")
    parser.add_argument("-f", "--folder", type=str, help="Destination folder for the split code")
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel: int) -> None:
    """Setup basic logging

    Parameters
    ----------
    loglevel : int
        minimum loglevel for emitting messages
    """
    log_format = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout, format=log_format, datefmt="%Y-%m-%d %H:%M:%S")


def main(args: List[str]) -> None:
    """Wrapper allowing :func:`split_code` to be called with string arguments in a CLI fashion

    Parameters
    ----------
    args : List[str])
        command line parameters as list of strings
        (for example  ``["-i", "my_source_code.py", "-f", "/path/to/output]``).
    """
    settings = parse_args(args=args)
    setup_logging(settings.loglevel)
    _logger.info(f"Split code file '{settings.input}' into folder '{settings.folder}'")
    split_code(settings.input, settings.folder)
    _logger.info("Script ends here")


def run() -> None:
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function is used as entry point for the console script by setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m code_split.code_split -i my_source_code.py
    #
    run()
